### PR TITLE
[WIP] feat Profile

### PR DESCRIPTION
**Pull request by TimotheeMathieu**
EDIT following comment by omar: 
This PR adds a method to agent_manager called `generate_profile`. Ot generates a python profile on one fit (otherwise python profiler does not work on parallel scripts by default). The profile is then saved in a file `profile.prof` and the 20 first lines are printed out.

I also added a message in case of profile generation that goes like this for for example on an experiment on Bandits:
```
[INFO] Doing a profile run. 
[INFO] Printing the 20 first lines of the profile 
         307412 function calls (307380 primitive calls) in 0.262 seconds

   Ordered by: cumulative time
   List reduced from 251 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.005    0.005    0.249    0.249 /USER/rlberry/rlberry/agents/bandits/index_agents.py:63(fit)
     2000    0.004    0.000    0.122    0.000 /USER/rlberry/rlberry/wrappers/writer_utils.py:28(step)
     2000    0.003    0.000    0.113    0.000 /USER/rlberry/rlberry/envs/bandits/bandit_base.py:34(step)
     2000    0.003    0.000    0.109    0.000 /USER/.local/lib/python3.10/site-packages/scipy/stats/_distn_infrastructure.py:470(rvs)
     2000    0.003    0.000    0.105    0.000 /USER/.local/lib/python3.10/site-packages/scipy/stats/_distn_infrastructure.py:3087(rvs)
     2000    0.018    0.000    0.102    0.000 /USER/.local/lib/python3.10/site-packages/scipy/stats/_distn_infrastructure.py:1032(rvs)
     1996    0.002    0.000    0.066    0.000 /USER/rlberry/rlberry/agents/bandits/indices.py:78(index)
     1996    0.029    0.000    0.064    0.000 /USER/rlberry/rlberry/agents/bandits/indices.py:79(<listcomp>)
     8023    0.006    0.000    0.058    0.000 {built-in method numpy.core._multiarray_umath.implement_array_function}
     2000    0.004    0.000    0.041    0.000 /USER/rlberry/rlberry/agents/bandits/tools/tracker.py:63(update)
     2000    0.007    0.000    0.030    0.000 /USER/.local/lib/python3.10/site-packages/scipy/stats/_distn_infrastructure.py:883(_argcheck_rvs)
     4000    0.002    0.000    0.029    0.000 <__array_function__ internals>:177(all)
    29952    0.026    0.000    0.026    0.000 /USER//rlberry/rlberry/utils/writers.py:100(read_last_tag_value)
     2004    0.001    0.000    0.025    0.000 /USER/rlberry/rlberry/utils/writers.py:182(add_scalars)
     2004    0.010    0.000    0.025    0.000 /USER/rlberry/rlberry/utils/writers.py:213(_add_scalars)
     1554    0.001    0.000    0.025    0.000 <string>:5(_parse_args_rvs)
    12009    0.018    0.000    0.024    0.000 /USER//rlberry/rlberry/utils/writers.py:140(_add_scalar)
     4000    0.003    0.000    0.023    0.000 /USER/.local/lib/python3.10/site-packages/numpy/core/fromnumeric.py:2404(all)
     4000    0.005    0.000    0.020    0.000 /USER/.local/lib/python3.10/site-packages/numpy/core/fromnumeric.py:69(_wrapreduction)
     2000    0.001    0.000    0.020    0.000 <__array_function__ internals>:177(broadcast_arrays)
[INFO] For more detailed info see the file profile.prof that has been created. 
```

Possible improvements : 
* preprocess the profile to have a column with the percentage of time spent instead of cumtime 
* generate a dot image instead of just a profile file (as in issue #189 ). This would require an additional dependency, maybe not worth it for now but may be useful later if we add features to the profile.
* investigate what happens when on GPU, i.e. how to use this for deep rl.
* do a profile for the memory (ram) used.
* add a profile to show where/when/how many time a seed is used/reseeded.
* add a profile to show when/where the environment is reset.
* If the fit is additive, we can also fit for several steps, monitor the state of the agent (with eval, looking at the weights of a NN...) and put these info into the profile.

More generally I think it would be nice to extend this feature to a more general profile. The idea is with just a change in arguments of the agent manager, I generate a file that contains information about what is under the hood to better understand how the training went.

Comments are welcome, in particular ideas about what could go into the profile and how to best format the profile are welcome.

### CODE REVIEW

**Comment by omardrwch** [comment]
Very nice idea! A suggestion: instead of sending `generate_profile` as a parameter to the constructor, I think it would be much better to have a method:
```
manager.generate_profile(fname)
```
that saves the profile to a file `fname`.


The advantages would be:
* No "hidden" extra fit() every time we call `manager.fit()`
* I think profiling is very useful to find and optimise the slow parts of the code, so it's nice to have a debugging phase where we run profiling (calling a `generate_profile` method) and, after that, we can train several instances with the `fit()` method.

**Comment by TimotheeMathieu** [comment]
Thanks, Ok I will make it a method. Remark that there was no hidden fit when generate_profile was false but I understand what you mean.

File: rlberry/manager/agent_manager.py [hunk 1]
```diff
@@ -568,6 +569,31 @@ def set_writer(self, idx, writer_fn, writer_kwargs=None):
         writer_kwargs = writer_kwargs or {}
         self.writers[idx] = (writer_fn, writer_kwargs)
 
+    def generate_profile(self, budget=None, fname="profile.prof"):
```
Comment (omardrwch) [inline rlberry/manager/agent_manager.py:572]: A last small suggestion: if `fname` is None (which could be the default), we can save it to `self.output_dir_ / "profile.prof"`

**Review by omardrwch — APPROVED** [review]

File: rlberry/manager/agent_manager.py [hunk 2]
```diff
@@ -568,6 +569,31 @@ def set_writer(self, idx, writer_fn, writer_kwargs=None):
...
+
+        fname: string, default="profile.prof"
+            name of the file where we save the profile.
+        """
+        budget = budget or self.fit_budget
+        logger.info("Doing a profile run.")
+        with cProfile.Profile() as pr:
+            agent = self.agent_class(**(self.init_kwargs[0]))
+            agent.fit(budget, **deepcopy(self.fit_kwargs))
+        pr.dump_stats("profile.prof")
```
[the remark this answers is not in the document]
Comment (TimotheeMathieu) [reply rlberry/manager/agent_manager.py:591]: I don't think so, this is straight from profiler doc (I rechecked) and I trust that python's dev were careful about this.

**Pull request merged**